src/lsm/forest_fuzz.zig

```diff
@@ -280,34 +315,58 @@ pub fn generate_fuzz_ops(random: std.rand.Random) ![]const FuzzOp {
+    var fuzz_op_distribution = fuzz.Distribution(FuzzOpTag){
+        // Maybe compact more often than forced to by `puts_since_compact`.
+        .compact = if (random.boolean()) 0 else 1,
+        // Always do puts, and always more puts than removes.
+        .put_account = config.lsm_batch_multiple * 2,
+        // Maybe do some gets.
+        .get_account = if (random.boolean()) 0 else config.lsm_batch_multiple,
+    };
     log.info("fuzz_op_distribution = {d:.2}", .{fuzz_op_distribution});
 
-    // We're not allowed to go more than Environment.cache_entries_max puts without compacting.
-    var puts_since_compact: usize = 0;
+    log.info("puts_since_compact_max = {}", .{Environment.puts_since_compact_max});
+    log.info("compacts_per_checkpoint = {}", .{Environment.compacts_per_checkpoint});
 
     var id_to_timestamp = std.hash_map.AutoHashMap(u128, u64).init(allocator);
     defer id_to_timestamp.deinit();
 
+    var op: u64 = 1;
+    var puts_since_compact: usize = 0;
     for (fuzz_ops) |*fuzz_op, fuzz_op_index| {
-        fuzz_op.* = if (puts_since_compact >= Environment.cache_entries_max)
+        const fuzz_op_tag: FuzzOpTag = if (puts_since_compact >= Environment.puts_since_compact_max)
             // We have to compact before doing any other operations.
-            FuzzOp{ .compact = {} }
+            .compact
         else
-        // Otherwise pick a random FuzzOp.
-        switch (fuzz.random_enum(random, std.meta.Tag(FuzzOp), fuzz_op_distribution)) {
-            .compact => FuzzOp{
-                .compact = {},
+            // Otherwise pick a random FuzzOp.
+            fuzz.random_enum(random, FuzzOpTag, fuzz_op_distribution);
+        fuzz_op.* = switch (fuzz_op_tag) {
+            .compact => compact: {
+                const compact_op = op;
+                op += 1;
+                const checkpoint =
+                    // Can only checkpoint on the last beat of the bar.
+                    compact_op % config.lsm_batch_multiple == config.lsm_batch_multiple - 1 and
+                    // Checkpoint at roughly the same rate as log wraparound.
+                    random.uintLessThan(usize, Environment.compacts_per_checkpoint) == 0;
+                break :compact FuzzOp{
+                    .compact = .{
+                        .op = compact_op,
+                        .checkpoint = checkpoint,
+                    },
+                };
             },
-            .put => put: {
+            .put_account => put_account: {
                 const id = random_id(random, u128);
                 // `timestamp` just needs to be unique, but we're not allowed to change the timestamp of an existing account.
                 const timestamp = id_to_timestamp.get(id) orelse fuzz_op_index;
                 try id_to_timestamp.put(id, timestamp);
-                break :put FuzzOp{ .put = Account{
+                break :put_account FuzzOp{ .put_account = Account{
```

Comment:
It could be inferred if you prefer. I've been defaulting to being explicit whenever the definition is not right next to the constraint that infers it.